nemo/collections/tts/modules/transformer_2501.py

```diff
@@ -0,0 +1,704 @@
+                torch.nn.init.normal_(param, mean=0.0, std=0.02 / math.sqrt(2 * n_layers))
+
+    def reset_cache(self, use_cache=False):
+        for layer in self.layers:
+            layer.reset_cache(use_cache)
+
+    @staticmethod
+    def _init_weights_gpt2(module):
+        if isinstance(module, (torch.nn.Linear, torch.nn.Embedding)):
+            torch.nn.init.normal_(module.weight, mean=0.0, std=0.02)
+
+        if isinstance(module, torch.nn.Linear) and module.bias is not None:
+            torch.nn.init.zeros_(module.bias)
+
+    @staticmethod
+    def _get_layer_inputs(
+        idx: int,
+        cond: Optional[Union[torch.Tensor, List[torch.Tensor]]],
+        cond_mask: Optional[Union[torch.Tensor, List[torch.Tensor]]],
+        attn_prior: Optional[Union[torch.Tensor, List[torch.Tensor]]],
+        multi_encoder_mapping: Optional[List[Optional[int]]],
+    ) -> Tuple[Optional[torch.Tensor], Optional[torch.Tensor], Optional[torch.Tensor]]:
+        if multi_encoder_mapping is not None:
+            if multi_encoder_mapping[idx] is None:
+                return None, None, None
+            else:
+                return (
+                    cond[multi_encoder_mapping[idx]],
+                    cond_mask[multi_encoder_mapping[idx]] if cond_mask is not None else None,
+                    attn_prior[multi_encoder_mapping[idx]] if attn_prior is not None else None,
+                )
+        else:
+            return cond, cond_mask, attn_prior
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        x_mask: torch.Tensor,
+        cond: Optional[Union[torch.Tensor, List[torch.Tensor]]] = None,
+        cond_mask: Optional[Union[torch.Tensor, List[torch.Tensor]]] = None,
+        attn_prior: Optional[Union[torch.Tensor, List[torch.Tensor]]] = None,
+        multi_encoder_mapping: Optional[List[Optional[int]]] = None,
+    ) -> Dict[str, Union[torch.Tensor, List]]:
+        """
+        Args:
+            x <torch tensor> (B, T1, C):
+            x_mask <bool mask> (B, T1): Multiplicative mask where True means we keep the input, False we zero it out.
+                Mostly used in non-causal self-attention to zero out padding values. In causal self-attention, the
+                causal mask will be used in place of this.
+            cond <torch tensor> (B, T2, C) or list of such tensors (from different encoders)
+            cond_mask <bool mask> (B, T2): Multiplicative mask where True means we keep the input, False we zero it
+                out or list of such tensors (from different encoders) output <torch tensor> (B, T1, C)
+            multi_encoder_mapping <list> <int>: None or Same size as n_layers, value indicates which cond input to use
+                for this layer
+
+        Returns dict with keys:
+            output <torch tensor> (B, T1, C): Output tensor
+            attn_probabilities <list>: Attention probabilities of each layer
+        """
```

Comment:
we should catch exceptions earlier here where the number of layers should be at least 2 when applying multiple conditions.

```suggestion
        """
        if isinstance(cond, list) and len(self.layers) < 2:
            raise ValueError(
                f"The number of Transformer layers should be no less than 2 when applying multiple conditionals. "
                f"Observed {len(self.layers)} layers on {len(cond)} conditionals."
            )
        
```